Remove LIBICONV_PLUG

krader1961 opened the pull request:
There was a period, roughly a year ago, when we needed to define
preprocessor symbol LIBICONV_PLUG on some platforms to build ksh. That
causes problems when building ksh via MacPorts. And testing on the
same platforms I was using a year ago when I introduced this hack shows
that it is apparently no longer needed. Presumably the real problem lay
elsewhere. Most likely in the various AST wrappers and `#define` symbols
that we've removed in the intervening time.

Resolves #932

siteshwar commented:
There is a reference to issue #280 about adding `LIBICONV_PLUG` symbol. AFAICT this change was done for macOS and gives priority to non-system version of `libiconv`.

@catull Are you able to build from this pull request on your system ?

krader1961 commented:
@siteshwar Note that #280 was merged by me with reservations and I had to subsequently amend the change to conditionally define that symbol. I have no trouble building and testing ksh with this change on macOS 10.12 and 10.13. Albeit without a non-standard libiconv installed.

ryandesign commented:
My understanding, from [the libiconv homepage](https://www.gnu.org/software/libiconv/) and from [this FreeBSD commit message](https://reviews.freebsd.org/rP398996) and other sources, is that `LIBICONV_PLUG` is not intended to be used by any software other than libiconv itself. It is only an option that may be used when compiling libiconv. Compiling libiconv with `LIBICONV_PLUG` means "don't use the `lib` prefix on `iconv` functions"; this "is a way to get good iconv support without having glibc-2.1". macOS doesn't use glibc; it uses libSystem, which doesn't contain any iconv functions. I assume Apple has used `LIBICONV_PLUG` when building the bundled copy of libiconv it ships in /usr. Compiling libiconv without `LIBICONV_PLUG` means "do use the `lib` prefix"; this is intended for most other libiconv installations and ensures that the libiconv functions don't interfere with the system's iconv functions. Whether libiconv is compiled with that option or not, as long as you `#include` the `<iconv.h>` matching the library you use, it should all work out, since the header will contain the right `#define`s so that you can always use the `iconv` functions without the `lib` prefix in your code.

krader1961 commented:
I checked the beta (93v-) branch and am pretty sure we introduced this hack because the code had stuff like this in *src/lib/libast/features/iconv*:

```
printf("#undef  iconv_open\n");
printf("#define iconv_open      %siconv_open\n", lib);
printf("extern %siconv_t        %siconv_open(const char*, const char*);\n", lib, lib);
```

That ended up dynamically constructing the *ast_iconv.h* header which contained entries like this which varies depending on whether or not the feature test determined that the  "lib" prefix was needed:

```
#undef      iconv_open
#define iconv_open  _ast_iconv_open
```

For expediency we generated the *ast_iconv.h* using the legacy IFFE mechanism one time on Linux then checked it into the project. It was only when I started working on getting ksh to build on other platforms that we found the problem with that `#define` above. Since then I removed that header as part of my removal of all the AST locale code that ksh didn't need or that we didn't want to support.

krader1961 commented:
Merged with commit comment amended.
